resources/hostname.go

```diff
@@ -0,0 +1,259 @@
+	bus, err := util.SystemBusPrivateUsable() // don't share the bus connection!
+	if err != nil {
+		return errwrap.Wrap(err, "Failed to connect to bus")
+	}
+	defer bus.Close()
+	bus.BusObject().Call("org.freedesktop.DBus.AddMatch", 0,
+		"type='signal',path='/org/freedesktop/hostname1',interface='org.freedesktop.DBus.Properties',member='PropertiesChanged'")
+	signals := make(chan *dbus.Signal, 10)
+	bus.Signal(signals)
+
+	var send = false // send event?
+	var exit = false
+	var dirty = false
+	var closed = false
+
+	defer func() {
+		if !closed {
+			close(signals)
+		}
+	}()
+
+	for {
+		obj.SetState(ResStateWatching) // reset
+		select {
+		case event := <-signals:
+			cuid.SetConverged(false)
+			if event == nil {
+				closed = true
+			}
+			send = true
+			dirty = true
+		case event := <-obj.Events():
+			cuid.SetConverged(false)
+			// we avoid sending events on unpause
+			if exit, send = obj.ReadEvent(&event); exit {
+				return nil // exit
+			}
+			send = true
+			dirty = true
+
+		case <-cuid.ConvergedTimer():
+			cuid.SetConverged(true) // converged!
+			continue
+
+		case <-Startup(startup):
+			cuid.SetConverged(false)
+			send = true
+		}
+
+		// do all our event sending all together to avoid duplicate msgs
+		if send {
+			startup = true // startup finished
+			send = false
+			if dirty {
+				dirty = false
+				obj.isStateOK = false // something made state dirty
+			}
+			// only do this on certain types of events
+			if exit, err := obj.DoSend(processChan, ""); exit || err != nil {
+				return err // we exit or bubble up a NACK...
+			}
+		}
+	}
+}
+
+func (obj *HostnameRes) updateProperty(object dbus.BusObject, property, setterName string) (bool, error) {
+	if hostnameProperty, err := object.GetProperty("org.freedesktop.hostname1." + property); err == nil {
+		switch hostname := hostnameProperty.Value().(type) {
+		case string:
+			if hostname != obj.Hostname {
+				log.Printf("Changing %s: %s => %s", property, hostname, obj.Hostname)
+				if err := object.Call("org.freedesktop.hostname1."+setterName, 0, obj.Hostname, false).Err; err != nil {
+					return false, errwrap.Wrapf(err, "failed to call org.freedesktop.hostname1.%s", setterName)
+				}
+				return true, nil
+			}
+		default:
+			return false, fmt.Errorf("Received unexpected type as %s value, expected string got '%t'", property, hostname)
+		}
+	} else {
+		return false, errwrap.Wrapf(err, "failed to get org.freedesktop.hostname1.%s", property)
+	}
+	return false, nil
+}
+
+// CheckApply method for Hostname resource. Does nothing, returns happy!
+func (obj *HostnameRes) CheckApply(apply bool) (checkok bool, err error) {
+	log.Printf("%v[%v]: CheckApply(%t)", obj.Kind(), obj.GetName(), apply)
+
+	conn, err := util.SystemBusPrivateUsable()
+	if err != nil {
+		return false, errwrap.Wrap(err, "Failed to connect to the private system bus")
+	}
+	defer conn.Close()
+
+	hostnameObject := conn.Object(hostname1Iface, hostname1Path)
```

Comment:
What's going on in the following 20 lines? I find it a bit confusing to read, and as a result, I'm lost.